File: src/main/java/com/werfad/finder/Finder.java
package com.werfad.finder;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.TextRange;
import com.werfad.MarksCanvas;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface Finder {

    /**
     * @return null - need more input to locate.
     * not null - can be locate some data, empty represent without any matches.
     */
    @Nullable
    List<MarksCanvas.Mark> start(@NotNull Editor e, @NotNull String s, @NotNull TextRange textRange);

    /**
     * @return same with {@link #start(Editor, String, TextRange)}
     */
    @Nullable
    List<MarksCanvas.Mark> input(@NotNull Editor e, char c, @NotNull List<MarksCanvas.Mark> lastMarks);
}

File: src/main/kotlin/com/werfad/finder/Finder.kt
package com.werfad.finder

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.util.TextRange
import com.werfad.MarksCanvas

interface Finder {
    /**
     * @return null - need more input to locate.
     *         not null - can be locate some data, empty represent without any matches.
     */
    fun start(e: Editor, s: String, visibleRange: TextRange): List<MarksCanvas.Mark>?

    /**
     * @return same with [.start]
     */
    fun input(e: Editor, c: Char, lastMarks: List<MarksCanvas.Mark>): List<MarksCanvas.Mark>?

    /**
     * @return Return the marks whose start character is removed.
     */
    fun advanceMarks(c: Char, marks: List<MarksCanvas.Mark>): List<MarksCanvas.Mark> {
        return marks.filter { it.keyTag[it.advanceIndex] == c }
            .map {
                MarksCanvas.Mark(it.keyTag, it.offset, it.advanceIndex + 1)
            }
            .toList()
    }
}